pkg/ingester/recalculate_owned_streams.go

```diff
@@ -2,32 +2,39 @@ package ingester
 
 import (
 	"context"
+	"fmt"
+	"sync"
 	"time"
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 	"github.com/grafana/dskit/ring"
 	"github.com/grafana/dskit/services"
+	"golang.org/x/exp/slices"
+
+	lokiring "github.com/grafana/loki/v3/pkg/util/ring"
 )
 
 type recalculateOwnedStreams struct {
 	services.Service
 
 	logger log.Logger
 
+	ownershipStrategy ownershipStrategy
 	instancesSupplier func() []*instance
-	ingesterID        string
-	previousRing      ring.ReplicationSet
-	ingestersRing     ring.ReadRing
 	ticker            *time.Ticker
 }
 
-func newRecalculateOwnedStreams(instancesSupplier func() []*instance, ingesterID string, ring ring.ReadRing, ringPollInterval time.Duration, logger log.Logger) *recalculateOwnedStreams {
+type ownershipStrategy interface {
+	checkRingForChanges() (bool, error)
+	isOwnedStream(*stream) (bool, error)
+}
+
+func newRecalculateOwnedStreamsSvc(instancesSupplier func() []*instance, ownershipStrategy ownershipStrategy, ringPollInterval time.Duration, logger log.Logger) *recalculateOwnedStreams {
```

Comment:
I think we should change both names, or none. For example `func newRecalculateOwnedStreamsSvc` and `type recalculateOwnedStreamsSvc`. Otherwise, it just requires a bit more effort to understand that this function creates a `recalculateOwnedStreams`.